### ISSUE TITLE
find_MAP will almost certainly be inaccurate for discrete stochastics

### ISSUE THREAD

**Issue by bjedwards** [pymc-devs/pymc3#396]
MAP estimates of the discrete stochastics will be inaccurate. Consider the following example:

``` python
import pymc as mc
with mc.Model() as model:
    p = mc.Beta('p',2,2)
    surv_sim = mc.Binomial('surv_sim',n=20,p=p)
    surv = mc.Binomial('surv',n=20,p=p,observed=15)
    map_est = mc.find_MAP(vars=model.vars)
print(map_est)
```

`map_est = {'p': array(0.5935401112367035), 'surv_sim': array(9.996758146781351)}`

Even if you specify that all variables be used (not the default), this is the case. Because discrete stochastics log probability functions return the log probability of the floor of any discrete stochastic(as far as I can tell), whichever scipy minimization function is used will be caught in a local optimum. The minimization will only optimize the continuous stochastics while keeping the discrete ones fixed at roughly the starting value as none of the small steps in the discrete variables results in an improvement of log probability.

I assume this is why the if `vars` is unspecified it defaults to `model.cont_vars`.

I am not sure if this is a bug or simply a fundamental limitation of finding MAP estimates when you have both discrete and continuous variables. I know in general integer programming is NP-Hard, but is there some special case we can exploit here? Feel free to close it if nothing can be done.

**Comment by twiecki** [comment]
Yeah, gradient based optimization will fail with discrete variables. You can try integrating the variables out but that's not always possible.

Another thought would be to apply global optimization methods (see e.g. hyperopt) to find a good starting point, this also works with discrete variables and also deal with multi-modal posteriors.

**Comment by bjedwards** [comment]
> Yeah, gradient based optimization will fail with discrete variables. You can try integrating the variables out but that's not always possible.

Speaking of gradient based optimization, I noticed that the `fmin` argument has to be a `scipy.optimize` function that accepts a gradient, otherwise you get a `TypeError`. Perhaps this should be a new issue, but being able to specify `fmin` or `fmin_powell` might be useful, and I think easily done (looking in `fmin.func_code.co_varnames` for `fprime`).

**Comment by jsalvatier** [comment]
That does sound useful. If you want to take a swing at it, I would gladly
accept a pull request.

**Comment by jsalvatier** [comment]
I think this is a fairly fundamental limitation for MAP finding. Any ideas
on how we could make this less surprising?

**Comment by bjedwards** [comment]
> That does sound useful. If you want to take a swing at it, I would gladly accept a pull request.

I'll fork and get started. I am unfamiliar with the pymc workflow, would you prefer a new issue be opened or can I just reference this one? Also I can't find any tests that deal specifically with find_MAP, but I could write a simple one or two.

**Comment by twiecki** [comment]
You can just reference this one but describe the intent in the PR.

Definitely find_MAP() tests would be highly welcome.

**Comment by bjedwards** [comment]
> I think this is a fairly fundamental limitation for MAP finding. Any ideas on how we could make this less surprising?

I'd advocate for raising a `Warning`. This should be easy to check for something like 

``` python
if any([v.type.dtype in mc.discrete_types for v in vars]):
    raise Warning("vars contains discrete stochastics, take any map with a grain of salt")
```

I would also advocate for changing the default `vars=model.vars`. That was my impression of what map was doing, until I looked at the function.

The suggestion to use something like `scipy.optimize.anneal` or `scipy.optimize.basinhopping` might be a good idea as these might be more successful and finding a global minimum.

**Comment by bjedwards** [comment]
This pull request changes `find_MAP` in a few ways. I changed my mind on providing the warning whenever discrete stochastics are present, and rather only do so when the user hasn't specified a new `fmin` function. I am making the (maybe bad) assumption that if an alternate function is provided the user realizes discrete stochastics might be a problem.

**Comment by bjedwards** [comment]
Fixed in #412 .
